# Nivell1
***Familiaritza't amb les tècniques d'exploració de les dades mitjantçant la estructura de dades, Dataframe amb la llibreria Pandas***

## Exercici 1
***Descarrega el data set Airlines Delay: Airline on-time statistics and delay causes i carrega’l a un pandas Dataframe. Explora les dades que conté, i queda’t únicament amb les columnes que consideris rellevants.***

In [110]:
#Importem les llibreries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
pd.options.display.max_columns=None
pd.options.display.max_rows=30

***Obrim el CSV separat alb comes,no hi ha un index i fem servir el encoding de defecte***


In [2]:
delayedFlightsDf=pd.read_csv('DelayedFlights.csv',index_col=0,sep=',', encoding='utf8')


C:\Users\Lenovo\anaconda3\lib\site-packages\numpy\lib\arraysetops.py:583: FutureWarning: elementwise comparison failed; returning scalar instead, but in the future will perform elementwise comparison
  mask |= (ar1 == a)


***Observem les columnes per decidir les que son significatives, max_colums=None***


In [3]:
delayedFlightsDf[:10]

Year  Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  ArrTime  \
0   2008      1           3          4   2003.0        1955   2211.0   
1   2008      1           3          4    754.0         735   1002.0   
2   2008      1           3          4    628.0         620    804.0   
4   2008      1           3          4   1829.0        1755   1959.0   
5   2008      1           3          4   1940.0        1915   2121.0   
6   2008      1           3          4   1937.0        1830   2037.0   
10  2008      1           3          4    706.0         700    916.0   
11  2008      1           3          4   1644.0        1510   1845.0   
15  2008      1           3          4   1029.0        1020   1021.0   
16  2008      1           3          4   1452.0        1425   1640.0   

    CRSArrTime UniqueCarrier  FlightNum TailNum  ActualElapsedTime  \
0         2225            WN        335  N712SW              128.0   
1         1000            WN       3231  N772SW              128.0   
2          750            WN        448  N428WN               96.0   
4         1925            WN       3920  N464WN               90.0   
5         2110            WN        378  N726SW              101.0   
6         1940            WN        509  N763SW              240.0   
10         915            WN        100  N690SW              130.0   
11        1725            WN       1333  N334SW              121.0   
15        1010            WN       2272  N263WN               52.0   
16        1625            WN        675  N286WN              228.0   

    CRSElapsedTime  AirTime  ArrDelay  DepDelay Origin Dest  Distance  TaxiIn  \
0            150.0    116.0     -14.0       8.0    IAD  TPA       810     4.0   
1            145.0    113.0       2.0      19.0    IAD  TPA       810     5.0   
2             90.0     76.0      14.0       8.0    IND  BWI       515     3.0   
4             90.0     77.0      34.0      34.0    IND  BWI       515     3.0   
5            115.0     87.0      11.0      25.0    IND  JAX       688     4.0   
6            250.0    230.0      57.0      67.0    IND  LAS      1591     3.0   
10           135.0    106.0       1.0       6.0    IND  MCO       828     5.0   
11           135.0    107.0      80.0      94.0    IND  MCO       828     6.0   
15            50.0     37.0      11.0       9.0    IND  MDW       162     6.0   
16           240.0    213.0      15.0      27.0    IND  PHX      1489     7.0   

    TaxiOut  Cancelled CancellationCode  Diverted  CarrierDelay  WeatherDelay  \
0       8.0          0                N         0           NaN           NaN   
1      10.0          0                N         0           NaN           NaN   
2      17.0          0                N         0           NaN           NaN   
4      10.0          0                N         0           2.0           0.0   
5      10.0          0                N         0           NaN           NaN   
6       7.0          0                N         0          10.0           0.0   
10     19.0          0                N         0           NaN           NaN   
11      8.0          0                N         0           8.0           0.0   
15      9.0          0                N         0           NaN           NaN   
16      8.0          0                N         0           3.0           0.0   

    NASDelay  SecurityDelay  LateAircraftDelay  
0        NaN            NaN                NaN  
1        NaN            NaN                NaN  
2        NaN            NaN                NaN  
4        0.0            0.0               32.0  
5        NaN            NaN                NaN  
6        0.0            0.0               47.0  
10       NaN            NaN                NaN  
11       0.0            0.0               72.0  
15       NaN            NaN                NaN  
16       0.0            0.0               12.0

***A partir del  describe veiem hi ha columnes contenen dades poc significatives SecurityDelay,CarrierDelay, etc***

In [4]:
delayedFlightsDf.describe()

Year         Month    DayofMonth     DayOfWeek       DepTime  \
count  1936758.0  1.936758e+06  1.936758e+06  1.936758e+06  1.936758e+06   
mean      2008.0  6.111106e+00  1.575347e+01  3.984827e+00  1.518534e+03   
std          0.0  3.482546e+00  8.776272e+00  1.995966e+00  4.504853e+02   
min       2008.0  1.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%       2008.0  3.000000e+00  8.000000e+00  2.000000e+00  1.203000e+03   
50%       2008.0  6.000000e+00  1.600000e+01  4.000000e+00  1.545000e+03   
75%       2008.0  9.000000e+00  2.300000e+01  6.000000e+00  1.900000e+03   
max       2008.0  1.200000e+01  3.100000e+01  7.000000e+00  2.400000e+03   

         CRSDepTime       ArrTime    CRSArrTime     FlightNum  \
count  1.936758e+06  1.929648e+06  1.936758e+06  1.936758e+06   
mean   1.467473e+03  1.610141e+03  1.634225e+03  2.184263e+03   
std    4.247668e+02  5.481781e+02  4.646347e+02  1.944702e+03   
min    0.000000e+00  1.000000e+00  0.000000e+00  1.000000e+00   
25%    1.135000e+03  1.316000e+03  1.325000e+03  6.100000e+02   
50%    1.510000e+03  1.715000e+03  1.705000e+03  1.543000e+03   
75%    1.815000e+03  2.030000e+03  2.014000e+03  3.422000e+03   
max    2.359000e+03  2.400000e+03  2.400000e+03  9.742000e+03   

       ActualElapsedTime  CRSElapsedTime       AirTime      ArrDelay  \
count       1.928371e+06    1.936560e+06  1.928371e+06  1.928371e+06   
mean        1.333059e+02    1.343027e+02  1.082771e+02  4.219988e+01   
std         7.206007e+01    7.134144e+01  6.864261e+01  5.678472e+01   
min         1.400000e+01   -2.500000e+01  0.000000e+00 -1.090000e+02   
25%         8.000000e+01    8.200000e+01  5.800000e+01  9.000000e+00   
50%         1.160000e+02    1.160000e+02  9.000000e+01  2.400000e+01   
75%         1.650000e+02    1.650000e+02  1.370000e+02  5.600000e+01   
max         1.114000e+03    6.600000e+02  1.091000e+03  2.461000e+03   

           DepDelay      Distance        TaxiIn       TaxiOut     Cancelled  \
count  1.936758e+06  1.936758e+06  1.929648e+06  1.936303e+06  1.936758e+06   
mean   4.318518e+01  7.656862e+02  6.812975e+00  1.823220e+01  3.268348e-04   
std    5.340250e+01  5.744797e+02  5.273595e+00  1.433853e+01  1.807562e-02   
min    6.000000e+00  1.100000e+01  0.000000e+00  0.000000e+00  0.000000e+00   
25%    1.200000e+01  3.380000e+02  4.000000e+00  1.000000e+01  0.000000e+00   
50%    2.400000e+01  6.060000e+02  6.000000e+00  1.400000e+01  0.000000e+00   
75%    5.300000e+01  9.980000e+02  8.000000e+00  2.100000e+01  0.000000e+00   
max    2.467000e+03  4.962000e+03  2.400000e+02  4.220000e+02  1.000000e+00   

           Diverted  CarrierDelay  WeatherDelay      NASDelay  SecurityDelay  \
count  1.936758e+06  1.247488e+06  1.247488e+06  1.247488e+06   1.247488e+06   
mean   4.003598e-03  1.917940e+01  3.703571e+00  1.502164e+01   9.013714e-02   
std    6.314722e-02  4.354621e+01  2.149290e+01  3.383305e+01   2.022714e+00   
min    0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   0.000000e+00   
25%    0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   0.000000e+00   
50%    0.000000e+00  2.000000e+00  0.000000e+00  2.000000e+00   0.000000e+00   
75%    0.000000e+00  2.100000e+01  0.000000e+00  1.500000e+01   0.000000e+00   
max    1.000000e+00  2.436000e+03  1.352000e+03  1.357000e+03   3.920000e+02   

       LateAircraftDelay  
count       1.247488e+06  
mean        2.529647e+01  
std         4.205486e+01  
min         0.000000e+00  
25%         0.000000e+00  
50%         8.000000e+00  
75%         3.300000e+01  
max         1.316000e+03

***Mostrem els noms de les columnes per identificar mes facilment les que volem eliminar***


In [5]:
print(delayedFlightsDf.columns.values)

['Year' 'Month' 'DayofMonth' 'DayOfWeek' 'DepTime' 'CRSDepTime' 'ArrTime'
 'CRSArrTime' 'UniqueCarrier' 'FlightNum' 'TailNum' 'ActualElapsedTime'
 'CRSElapsedTime' 'AirTime' 'ArrDelay' 'DepDelay' 'Origin' 'Dest'
 'Distance' 'TaxiIn' 'TaxiOut' 'Cancelled' 'CancellationCode' 'Diverted'
 'CarrierDelay' 'WeatherDelay' 'NASDelay' 'SecurityDelay'
 'LateAircraftDelay']


***Eliminem les columnes que no estem ineteressats, mantenim el mateix Dataframe***

In [6]:
delayedFlightsDf.drop(['TailNum','CarrierDelay','CancellationCode','Diverted','CarrierDelay','WeatherDelay','NASDelay',
 'SecurityDelay' ,'LateAircraftDelay','CRSElapsedTime','CRSArrTime','DepDelay','TaxiIn' ,'TaxiOut'],axis=1,inplace=True)

In [7]:
delayedFlightsDf[:10]

Year  Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  ArrTime  \
0   2008      1           3          4   2003.0        1955   2211.0   
1   2008      1           3          4    754.0         735   1002.0   
2   2008      1           3          4    628.0         620    804.0   
4   2008      1           3          4   1829.0        1755   1959.0   
5   2008      1           3          4   1940.0        1915   2121.0   
6   2008      1           3          4   1937.0        1830   2037.0   
10  2008      1           3          4    706.0         700    916.0   
11  2008      1           3          4   1644.0        1510   1845.0   
15  2008      1           3          4   1029.0        1020   1021.0   
16  2008      1           3          4   1452.0        1425   1640.0   

   UniqueCarrier  FlightNum  ActualElapsedTime  AirTime  ArrDelay Origin Dest  \
0             WN        335              128.0    116.0     -14.0    IAD  TPA   
1             WN       3231              128.0    113.0       2.0    IAD  TPA   
2             WN        448               96.0     76.0      14.0    IND  BWI   
4             WN       3920               90.0     77.0      34.0    IND  BWI   
5             WN        378              101.0     87.0      11.0    IND  JAX   
6             WN        509              240.0    230.0      57.0    IND  LAS   
10            WN        100              130.0    106.0       1.0    IND  MCO   
11            WN       1333              121.0    107.0      80.0    IND  MCO   
15            WN       2272               52.0     37.0      11.0    IND  MDW   
16            WN        675              228.0    213.0      15.0    IND  PHX   

    Distance  Cancelled  
0        810          0  
1        810          0  
2        515          0  
4        515          0  
5        688          0  
6       1591          0  
10       828          0  
11       828          0  
15       162          0  
16      1489          0

***Observem tipus de dades , semblen massa generics***

In [8]:
delayedFlightsDf.dtypes

Year                   int64
Month                  int64
DayofMonth             int64
DayOfWeek              int64
DepTime              float64
CRSDepTime             int64
ArrTime              float64
UniqueCarrier         object
FlightNum              int64
ActualElapsedTime    float64
AirTime              float64
ArrDelay             float64
Origin                object
Dest                  object
Distance               int64
Cancelled              int64
dtype: object

## Exercici 2
***Fes un informe complet del data set:.***

***Resumeix estadísticament les columnes d’interès***
* Troba quantes dades faltants hi ha per columna
* Crea columnes noves (velocitat mitjana del vol, si ha arribat tard o no...)
* Taula de les aerolínies amb més endarreriments acumulats
* Quins són els vols més llargs? I els més endarrerits?
* Etc.

### Resum estadistic de columnes

In [9]:
delayedFlightsDf.describe()

Year         Month    DayofMonth     DayOfWeek       DepTime  \
count  1936758.0  1.936758e+06  1.936758e+06  1.936758e+06  1.936758e+06   
mean      2008.0  6.111106e+00  1.575347e+01  3.984827e+00  1.518534e+03   
std          0.0  3.482546e+00  8.776272e+00  1.995966e+00  4.504853e+02   
min       2008.0  1.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%       2008.0  3.000000e+00  8.000000e+00  2.000000e+00  1.203000e+03   
50%       2008.0  6.000000e+00  1.600000e+01  4.000000e+00  1.545000e+03   
75%       2008.0  9.000000e+00  2.300000e+01  6.000000e+00  1.900000e+03   
max       2008.0  1.200000e+01  3.100000e+01  7.000000e+00  2.400000e+03   

         CRSDepTime       ArrTime     FlightNum  ActualElapsedTime  \
count  1.936758e+06  1.929648e+06  1.936758e+06       1.928371e+06   
mean   1.467473e+03  1.610141e+03  2.184263e+03       1.333059e+02   
std    4.247668e+02  5.481781e+02  1.944702e+03       7.206007e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00       1.400000e+01   
25%    1.135000e+03  1.316000e+03  6.100000e+02       8.000000e+01   
50%    1.510000e+03  1.715000e+03  1.543000e+03       1.160000e+02   
75%    1.815000e+03  2.030000e+03  3.422000e+03       1.650000e+02   
max    2.359000e+03  2.400000e+03  9.742000e+03       1.114000e+03   

            AirTime      ArrDelay      Distance     Cancelled  
count  1.928371e+06  1.928371e+06  1.936758e+06  1.936758e+06  
mean   1.082771e+02  4.219988e+01  7.656862e+02  3.268348e-04  
std    6.864261e+01  5.678472e+01  5.744797e+02  1.807562e-02  
min    0.000000e+00 -1.090000e+02  1.100000e+01  0.000000e+00  
25%    5.800000e+01  9.000000e+00  3.380000e+02  0.000000e+00  
50%    9.000000e+01  2.400000e+01  6.060000e+02  0.000000e+00  
75%    1.370000e+02  5.600000e+01  9.980000e+02  0.000000e+00  
max    1.091000e+03  2.461000e+03  4.962000e+03  1.000000e+00

#### Valors nuls per columna

In [10]:
delayedFlightsDf.isnull().sum(axis=0)

Year                    0
Month                   0
DayofMonth              0
DayOfWeek               0
DepTime                 0
CRSDepTime              0
ArrTime              7110
UniqueCarrier           0
FlightNum               0
ActualElapsedTime    8387
AirTime              8387
ArrDelay             8387
Origin                  0
Dest                    0
Distance                0
Cancelled               0
dtype: int64

#### Mirem les dades de valors nuls

In [11]:
delayedFlightsDf[delayedFlightsDf.ActualElapsedTime.isnull()]

Year  Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  ArrTime  \
1763     2008      1           3          4    922.0         915      NaN   
1911     2008      1           3          4   2325.0        1900      NaN   
2651     2008      1           4          5   1949.0        1905      NaN   
2726     2008      1           4          5    737.0         705      NaN   
3672     2008      1           4          5   1849.0        1630      NaN   
...       ...    ...         ...        ...      ...         ...      ...   
7006289  2008     12          10          3   1459.0        1447      NaN   
7006401  2008     12          11          4   1355.0        1106      7.0   
7006809  2008     12          11          4   1026.0         955      NaN   
7007034  2008     12          11          4   1527.0        1520   2106.0   
7008584  2008     12          12          5    703.0         630      NaN   

        UniqueCarrier  FlightNum  ActualElapsedTime  AirTime  ArrDelay Origin  \
1763               WN       1069                NaN      NaN       NaN    SAN   
1911               WN       2092                NaN      NaN       NaN    SFO   
2651               WN       1403                NaN      NaN       NaN    BOI   
2726               WN        178                NaN      NaN       NaN    BUR   
3672               WN        239                NaN      NaN       NaN    LAS   
...               ...        ...                ...      ...       ...    ...   
7006289            DL       1706                NaN      NaN       NaN    ATL   
7006401            DL         26                NaN      NaN       NaN    LAX   
7006809            DL        892                NaN      NaN       NaN    ATL   
7007034            DL       1102                NaN      NaN       NaN    IAD   
7008584            DL       1372                NaN      NaN       NaN    LGA   

        Dest  Distance  Cancelled  
1763     SMF       480          0  
1911     SAN       447          0  
2651     RNO       335          0  
2726     SMF       358          0  
3672     RNO       345          0  
...      ...       ...        ...  
7006289  BUF       712          1  
7006401  JFK      2475          0  
7006809  JFK       760          1  
7007034  ATL       533          0  
7008584  BOS       185          1  

[8387 rows x 16 columns]

#### Eliminem valors Nan i verifiquem que ja no en queden

In [14]:
delayedFlightsDf.dropna(inplace=True)

In [15]:
delayedFlightsDf[delayedFlightsDf.ActualElapsedTime.isnull()]

Empty DataFrame
Columns: [Year, Month, DayofMonth, DayOfWeek, DepTime, CRSDepTime, ArrTime, UniqueCarrier, FlightNum, ActualElapsedTime, AirTime, ArrDelay, Origin, Dest, Distance, Cancelled]
Index: []

### Afegir noves columnes y neteja adicional

#### Afegir velocitat mitjana total: AvgSpeed 

In [16]:
delayedFlightsDf['AvgSpeed']=delayedFlightsDf['Distance']*60/delayedFlightsDf['ActualElapsedTime']

In [60]:
delayedFlightsDf['AvgSpeed']

0          379.687500
1          379.687500
2          321.875000
4          343.333333
5          408.712871
              ...    
7009710    369.795918
7009717    227.244094
7009718    255.185185
7009726    278.086957
7009727    426.341463
Name: AvgSpeed, Length: 1928371, dtype: float64


#### Afegir velocitat mitjana en vol:  AvgSpeedFlight

In [ ]:
delayedFlightsDf['AvgSpeedFlight']=delayedFlightsDf['Distance']*60/delayedFlightsDf['AirTime']

In [64]:
delayedFlightsDf['AvgSpeedFlight']

0          418.965517
1          430.088496
2          406.578947
4          401.298701
5          474.482759
              ...    
7009710    453.000000
7009717    370.000000
7009718    338.852459
7009726    359.325843
7009727    504.230769
Name: AvgSpeedFlight, Length: 1928371, dtype: float64

#### Vols amb retard True/False-Delayed

In [18]:
delayedDf=[]
dalayed =False
for delay in delayedFlightsDf['ArrDelay']:
    if delay >0:
        delayed=True
    else:
        delayed=False
    delayedDf.append(delayed)
   

In [19]:
delayedFlightsDf['Delayed']=delayedDf


#### Comprovem les dues columnes afegides***

In [109]:
delayedFlightsDf[:10]

Year  Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  ArrTime  \
0   2008      1           3          4   2003.0        1955   2211.0   
1   2008      1           3          4    754.0         735   1002.0   
2   2008      1           3          4    628.0         620    804.0   
4   2008      1           3          4   1829.0        1755   1959.0   
5   2008      1           3          4   1940.0        1915   2121.0   
6   2008      1           3          4   1937.0        1830   2037.0   
10  2008      1           3          4    706.0         700    916.0   
11  2008      1           3          4   1644.0        1510   1845.0   
15  2008      1           3          4   1029.0        1020   1021.0   
16  2008      1           3          4   1452.0        1425   1640.0   

   UniqueCarrier  FlightNum  ActualElapsedTime  AirTime  ArrDelay Origin Dest  \
0             WN        335              128.0    116.0     -14.0    IAD  TPA   
1             WN       3231              128.0    113.0       2.0    IAD  TPA   
2             WN        448               96.0     76.0      14.0    IND  BWI   
4             WN       3920               90.0     77.0      34.0    IND  BWI   
5             WN        378              101.0     87.0      11.0    IND  JAX   
6             WN        509              240.0    230.0      57.0    IND  LAS   
10            WN        100              130.0    106.0       1.0    IND  MCO   
11            WN       1333              121.0    107.0      80.0    IND  MCO   
15            WN       2272               52.0     37.0      11.0    IND  MDW   
16            WN        675              228.0    213.0      15.0    IND  PHX   

    Distance  Cancelled    AvgSpeed  Delayed  AvgSpeedFlight  
0        810          0  379.687500    False      418.965517  
1        810          0  379.687500     True      430.088496  
2        515          0  321.875000     True      406.578947  
4        515          0  343.333333     True      401.298701  
5        688          0  408.712871     True      474.482759  
6       1591          0  397.750000     True      415.043478  
10       828          0  382.153846     True      468.679245  
11       828          0  410.578512     True      464.299065  
15       162          0  186.923077     True      262.702703  
16      1489          0  391.842105     True      419.436620

***Checking de velocitats , velocitats <180 0 > 700 mph np son realistas***

In [77]:
delayedFlightsDf[(delayedFlightsDf.AvgSpeedFlight>700) | (delayedFlightsDf.AvgSpeedFlight<180)].loc[:,['AirTime', 'Distance','AvgSpeedFlight']].sort_values('Distance',ascending=False)

AirTime  Distance  AvgSpeedFlight
5180146   1091.0      2677      147.222731
1118326    214.0      2569      720.280374
2092895     83.0      2248     1625.060241
1768574    167.0      1968      707.065868
4180520    130.0      1597      737.076923
...          ...       ...             ...
585919      11.0        30      163.636364
151660      11.0        24      130.909091
1637250     10.0        24      144.000000
2547298     31.0        11       21.290323
4392215     14.0        11       47.142857

[10885 rows x 3 columns]

***Eliminem els vols de velocitats no coherents 10885 files***

In [79]:
delayedFlightsDf.drop(delayedFlightsDf[((delayedFlightsDf['AvgSpeedFlight']>700) | (delayedFlightsDf['AvgSpeedFlight']<180))].index,inplace=True)

### Consultes varies


#### companyies en major retard acumulat

In [113]:
delayedFlightsDf.loc[:,['UniqueCarrier','ArrDelay']].groupby('UniqueCarrier').sum().sort_values('ArrDelay',ascending=False).head(5)

ArrDelay
UniqueCarrier            
WN             11279776.0
AA              8876420.0
UA              6721559.0
MQ              6266705.0
OO              5751242.0

#### Vols per companyia

In [114]:
delayedFlightsDf.groupby(['UniqueCarrier']).size().reset_index(name='counts').sort_values('counts',ascending=False).head(10)

UniqueCarrier  counts
17            WN  375883
1             AA  190800
15            UA  140797
11            MQ  139688
14            OO  128481
6             DL  113577
18            XE  102537
5             CO   99458
16            US   97641
7             EV   81058

#### Vols mes llargs agrupament per Origin i Desti-Dades

In [81]:
delayedFlightsDf.loc[:,['Origin','Dest','Distance']].groupby(['Origin','Dest']).max().sort_values('Distance',ascending=False).head(10)

Distance
Origin Dest          
EWR    HNL       4962
HNL    EWR       4962
       ATL       4502
ATL    HNL       4502
HNL    ORD       4243
ORD    HNL       4243
KOA    ORD       4213
ORD    OGG       4184
MSP    HNL       3972
HNL    MSP       3972

#### Vols mes curst agrupament per Origin i Desti-Dades

In [94]:
delayedFlightsDf.loc[:,['Origin','Dest','Distance']].groupby(['Origin','Dest']).max().sort_values('Distance',ascending=True).head(20)

Distance
Origin Dest          
WRG    PSG         31
PSG    WRG         31
HPN    EWR         35
LAX    SNA         36
PUB    COS         37
SRQ    TPA         40
GST    JNU         41
JNU    GST         41
MKG    GRR         41
ONT    LAX         47
OXR    LAX         49
LAX    OXR         49
DTW    FNT         56
ACV    CEC         56
CEC    ACV         56
FNT    DTW         56
YUM    IPL         58
IPL    YUM         58
CVG    DAY         64
DAY    CVG         64

#### Vols mes ràpids

In [84]:
delayedFlightsDf.loc[:,['Origin','Dest','AvgSpeedFlight']].groupby(['Origin','Dest']).max().sort_values('AvgSpeedFlight',ascending=False).head(20)

AvgSpeedFlight
Origin Dest                
CVG    LEX       700.000000
PHX    SNA       699.310345
HNL    KOA       698.571429
ASE    SLC       698.400000
CLT    MEM       698.181818
ATL    DFW       697.142857
LIT    ATL       696.923077
CVG    EWR       696.734694
HPN    MCO       694.285714
JFK    MSP       693.033708
SLC    AUS       692.553191
SAT    IAD       692.542373
ORD    BUF       692.195122
MCI    SMF       692.160000
SEA    PHX       691.875000
DEN    COD       691.764706
CLT    BDL       690.000000
ICT    ATL       689.117647
DFW    ELP       688.750000
AUS    ATL       687.042254

#### Mediana I media  de la velocitat dels vols dona un valor mes aproximat

In [86]:
delayedFlightsDf['AvgSpeedFlight'].median()

404.375

In [95]:
delayedFlightsDf['AvgSpeedFlight'].mean()

398.06250611232065

## Exercici 3
***Exporta el data set net i amb les noves columnes a CSV complert i a excel filtrat per companyia WN/AA Excel.***

In [104]:
delayedFlightsDf.to_csv('DelayedFlights_net.csv')

In [108]:
delayedFlightsDf[((delayedFlightsDf.UniqueCarrier=='WN')| (delayedFlightsDf.UniqueCarrier=='AA'))].to_excel('DelayedFlights_net.xlsx')